### PR TITLE
Refactor two-body state vector representation and expose from Python

### PR DESCRIPTION
**Pull request by helgee**
This PR accomplishes two goals:

- Make the two-body state vector representations (`Cartesian` and `Keplerian`) usable from Python
- Attach a reference frame to the two-body state vector representations

## Two-Body Python Classes

The main problem here is that [PyO3 does not support generics](https://pyo3.rs/v0.20.2/class#restrictions). My solution is to expose a subset of the functionality which is provided via traits and generics on the Rust side as enums (see `PyBody` and `PyFrame`) and split the representations into a concrete core (e.g. `CartesianState`) and a generic wrapper (e.g. `Cartesian`) so that the concrete core can be reused in the Python class (e.g. `PyCartesian`).

## Reference Frames

The generic wrapper implements the new `CoordinateSystem` trait which is parametrized by the central body/coordinate origin and the frame of reference. The next step is to implement reference frame transformation methods.

## To Do

- [x] Improve test coverage
- [x] Add Python type hints

### CODE REVIEW

File: crates/lox_core/src/two_body.rs [hunk 1]
```diff
@@ -33,189 +193,239 @@ pub trait TwoBody {
...
-    epoch: Epoch,
-    center: T,
-    position: DVec3,
-    velocity: DVec3,
+pub struct Cartesian<T, S>
+where
+    T: PointMass + Copy,
+    S: ReferenceFrame + Copy,
+{
+    time: Epoch,
```
Comment (helgee) [inline crates/lox_core/src/two_body.rs:201]: The time field should probably be part of the concrete struct.

**Review by matzipan — COMMENTED** [review]
I sort of follow along but I can't say I fully grasp this MR. That is to say, I don't see anything wrong with it.

File: crates/lox_core/src/frames.rs [hunk 2]
```diff
@@ -15,11 +15,25 @@ pub mod iau;
 // TODO: Replace with proper `Epoch` type
 type Epoch = f64;
 
-pub trait ReferenceFrame {
-    fn is_inertial(&self) -> bool;
+pub trait ReferenceFrame {}
+
+pub trait InertialFrame: ReferenceFrame {
+    fn is_inertial(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (helgee) [reply crates/lox_core/src/frames.rs:21]: No, but then we'd lose object safety.

File: crates/lox_core/src/two_body.rs [hunk 3]
```diff
@@ -9,21 +9,181 @@
...
+            eccentricity,
+            inclination,
+            ascending_node,
+            periapsis_arg,
+            true_anomaly,
+        }
+    }
+
+    pub fn semi_major(&self, grav_param: f64) -> f64 {
+        self.to_keplerian_state(grav_param).semi_major
```
[the remark this answers is not in the document]
Comment (helgee) [reply crates/lox_core/src/two_body.rs]: Made the conversion explicit.

File: crates/lox_core/src/two_body.rs [hunk 4]
```diff
@@ -33,189 +193,239 @@ pub trait TwoBody {
...
 
 #[derive(Debug, Clone)]
-pub struct Cartesian<T: PointMass + Copy> {
-    epoch: Epoch,
-    center: T,
-    position: DVec3,
-    velocity: DVec3,
+pub struct Cartesian<T, S>
+where
+    T: PointMass + Copy,
```
[the remark this answers is not in the document]
Comment (helgee) [reply crates/lox_core/src/two_body.rs]: The compiler does not know that all types that implement these traits are ZST and thus implicitly copyable. Without this requirement the borrow checker would complain in the conversion methods.

File: crates/lox_py/src/time.rs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+use crate::LoxPyError;
+use lox_core::time::dates::{Date, Time};
+use lox_core::time::epochs::{Epoch, TimeScale};
+use pyo3::{pyclass, pymethods};
+
+#[pyclass(name = "TimeScale")]
+pub struct PyTimeScale(pub TimeScale);
+
+#[pymethods]
+impl PyTimeScale {
```
[the remark this answers is not in the document]
Comment (helgee) [reply crates/lox_py/src/time.rs:18]: I plan to rewrite the Python wrapper for the `time` module once you are done. This was my first experiment with PyO3.

File: crates/lox_py/src/two_body.rs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+    fn ascending_node(&self) -> f64 {
+        self.state.ascending_node()
+    }
+
+    fn periapsis_arg(&self) -> f64 {
+        self.state.periapsis_arg()
+    }
+
+    fn true_anomaly(&self) -> f64 {
```
[the remark this answers is not in the document]
Comment (helgee) [reply crates/lox_py/src/two_body.rs]: I'll try to come up with a way to test from Python. It does not seem to be possible from Rust (due to a dependency on the NumPy C-API).

[the remark this answers is not in the document]
Comment (helgee) [reply crates/lox_py/src/two_body.rs]: ...in a subsequent PR.

**Comment by helgee** [comment]
Ready for another round of reviews.

File: crates/lox_core/src/two_body.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (helgee) [reply crates/lox_core/src/two_body.rs]: I originally added Copy as a supertrait (also for the other traits) but this had unpleasant knock-on effects. Unfortunately, I do not remember what the problem was exactly.

**Pull request merged**